File: src/merge_DGE_PRED.py


#import libraries
import argparse
import pandas as pd
import glob


#Argunent Parser
#**************
parser = argparse.ArgumentParser(description='Merge APA prediction with DGE Matrix')
parser.add_argument('PRED_path', metavar='pfip', type=str, help='path of cell isoform prediction file')
parser.add_argument('DGE_path', metavar='dfip', type=str, help='path of DGE file')
parser.add_argument('OUTPUT_APADGE_path', metavar='opfop', type=str, help='path of output APA_DGE file')
args = parser.parse_args()


#Open files
print('Files opening...')
predictions = pd.read_csv(args.PRED_path, names = ["bc","gene_name","transcript_name","tr_prob"], sep="\t", skiprows = 1)

print('dge...')
dge = pd.read_csv(args.DGE_path, sep="\t")
dge = dge.rename(columns = {'GENE':'gene_name'})

#melt dge table
print('melting table...')
dge = dge.melt(id_vars='gene_name')

predictions["bc_gn"] = predictions.bc + "_" + predictions.gene_name
dge["bc_gn"] = dge.variable + "_" + dge.gene_name

#filter barcodes not present in the dge
print('bc filtering..')
dge = dge[dge.bc_gn.isin(predictions.bc_gn.tolist())]
dge = dge.rename(columns = {'variable':'bc', 'value':'counts'})

#merge with predictions table
dge = predictions[['bc','gene_name','transcript_name','tr_prob']].merge(dge, on=['bc','gene_name'], how='left')

#Get relatives values on dge
dge['gene_transcript'] = dge.gene_name + '***' + dge.transcript_name
dge['relative_prob'] = dge.counts * dge.tr_prob
dge = dge[['bc','gene_transcript','relative_prob']]

#pivot of table
print('table pivoting...')
dge = dge.pivot_table(index='gene_transcript', columns='bc', values='relative_prob', fill_value=0)

#writing
print('Writing...')
dge.to_csv(args.OUTPUT_APADGE_path, sep="\t")

File: src/merge_DGE_PRED.R
library(dplyr)
library(data.table)
library(argparse)
library(stringr)
library(tidyr)


#Argument Parser
#===============
parser = ArgumentParser(description='Merging DGE & isoforms Prediction')
parser$add_argument('DGE', type="character", help='path of DGE file')
parser$add_argument('PRED', type="character", help='path of predicted isoforms')
parser$add_argument('OUTP', type="character", help="path of output file")
args = parser$parse_args()

#Files opening
preds = fread(args$PRED, col.names=c("bc","gene_name","transcript_name","tr_prob"))
dge = fread(args$DGE)

#melting
dge = melt(dge)

#renaming
colnames(dge) = c("gene_name","bc","counts")

#left_joining by barcodes & gene_name
dge_preds = left_join(preds, dge) %>% stats::na.omit() %>% mutate(tr_counts=tr_prob * counts, gene_transcript = paste0(gene_name, "***", transcript_name)) %>% distinct(gene_transcript, bc, tr_counts) %>% pivot_wider(names_from=bc, values_from=tr_counts, values_fill = 0)


#writing
fwrite(dge_preds, file=args$OUTP, sep="\t")
